src/trfs_fast/llama.py

```diff
@@ -104,44 +104,55 @@ def __init__(self, dim, max_position_embeddings=2048, base=10000, device=None):
-        )
+    # def forward(self, x, seq_len=None):
+    #     # x: [bs, num_attention_heads, seq_len, head_size]
+    #     # This `if` block is unlikely to be run after we build sin/cos in `__init__`. Keep the logic here just in case.
+
+    #     # raises control flow exception
+    #     # if seq_len > self.max_seq_len_cached:
+    #     #     self.max_seq_len_cached = seq_len
+    #     #     t = torch.arange(self.max_seq_len_cached, device=x.device, dtype=self.inv_freq.dtype)
+    #     #     freqs = torch.einsum("i,j->ij", t, self.inv_freq)
+    #     #     # Different from paper, but it uses a different permutation in order to obtain the same calculation
+    #     #     emb = torch.cat((freqs, freqs), dim=-1).to(x.device)
+    #     #     self.register_buffer("cos_cached", emb.cos()[None, None, :, :].to(x.dtype), persistent=False)
+    #     #     self.register_buffer("sin_cached", emb.sin()[None, None, :, :].to(x.dtype), persistent=False)
 
+    #     return (
+    #         self.cos_cached.to(dtype=x.dtype),
+    #         self.sin_cached.to(dtype=x.dtype),
+    #     )
 
-def rotate_half(x):
-    """Rotates half the hidden dims of the input."""
-    x = torch.roll(x, shifts=x.shape[-1] // 2, dims=-1)
-    x[..., : x.shape[-1] // 2] *= -1
-    return x
 
+# def rotate_half(x):
+#     """Rotates half the hidden dims of the input."""
+#     x = torch.roll(x, shifts=x.shape[-1] // 2, dims=-1)
+#     x[..., : x.shape[-1] // 2] *= -1
+#     return x
 
 
-def apply_rotary_pos_emb_opt(q, key_states, cos, sin, position_ids):
+
+# def apply_rotary_pos_emb_opt(q, key_states, cos, sin, position_ids):
+def apply_rotary_pos_emb_opt(q, key_states, cos, sin):
     # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
     # TODO: can we remove some squeeze/unsqueeze?
     # TODO: replace by squeeze(0, 1) once https://github.com/pytorch/pytorch/issues/103875 is fixed
-    cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
-    sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
-    cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
-    sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
-    q_embed = (q * cos) + (rotate_half(q) * sin)
-    key_states.copy_((key_states * cos) + (rotate_half(key_states) * sin))
+    # cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+    # sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+    # cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    # sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+    half_dim = cos.shape[-1] // 2
+    q_embed = (q * cos)
+    q_embed[..., half_dim:] += q[..., :half_dim] * sin[..., half_dim:]
+    q_embed[..., :half_dim] += q[..., half_dim:] * sin[..., :half_dim] * -1
+
+    k_embed = (key_states * cos)
+    k_embed[..., half_dim:] += key_states[..., :half_dim] * sin[..., half_dim:]
+    k_embed[..., :half_dim] += key_states[..., half_dim:] * sin[..., :half_dim] * -1
```

Comment:
I assume there is a lot of overhead from multiple aten::slice calls

Reply:
The alternative, calling `rotate_half`, is equally as bad 😅 

It's quite frustrating to know that our attention layers take more time to apply the rotary embedding than the attention itself 😞

Reply:
Yes, at least it's good to know there's a strong bottleneck there. Maybe there exist better pytorch-based implementation (not sure how TGI handles it).